Implement metrics for external queue

Atharva-Kanherkar opened the pull request:
## WorkStealingThreadPool Metrics Implementation

This PR implements metrics for tracking singleton and batch task submissions to the external queue in `WorkStealingThreadPool`, as  in issue #4269 .

### Changes Made:

1. **Added Atomic Counters to WorkStealingThreadPool**:
   - `singletonsSubmittedCount` - tracks total singleton task submissions
   - `singletonsPresentCount` - tracks current singleton tasks in the queue
   - `batchesSubmittedCount` - tracks total batch task submissions
   - `batchesPresentCount` - tracks current batch tasks in the queue

2. **Updated Submission Logic**:
   - Changed `submitToExternalQueue()` to increment counters when tasks are added
   - Changed `pollTask()` to decrement present counters when tasks are polled

3. **Added Accessor Methods**:
   - `getSingletonsSubmittedCount()` - returns total singleton submissions 
   - `getSingletonsPresentCount()` - returns current singletons in queue
   - `getBatchesSubmittedCount()` -  returns total batch submissions
   - `getBatchesPresentCount()` -  returns current batches in queue
   - `logQueueMetrics()` - Helper for logging all metrics

4. **Added Verification Test**:
    I made a test (which i am not sure about, but It did work -->)
![image](https://github.com/user-attachments/assets/74f05823-4809-4592-b788-2750ba43b687)

    - Submits singleton tasks and verifies the counter increases
   - Triggers batch creation through local queue overflow
   - Verifies both types of metrics track correctly

- Singleton submissions: Created 10,000 tasks → Counter increased by exactly 10,000 ✓
- Batch submissions: Generated queue overflow → Counter increased by 2,053 ✓

This might confirm that the implementation works as fine as it should be. The code too compiles without any errors, at least  in my terminal :)
The test is included in the PR for reference, showing how to access these metrics from client code.

Atharva-Kanherkar committed:
Adding metrics

Atharva-Kanherkar committed:
Added demo test and some other getter methods

Atharva-Kanherkar committed:
Removed the unnecessary commit

Atharva-Kanherkar committed:
removed unnecessary files

Atharva-Kanherkar committed:
Remove .scala-build/ files from repository

Atharva-Kanherkar committed:
Add metrics for external queue tracking in WorkStealingThreadPool

Atharva-Kanherkar committed:
Final commit

Atharva-Kanherkar committed:
Some changes

Atharva-Kanherkar commented:
Hey @armanbilge! Please ignore some of the previous commits as i got myself into a git command hell :p , Anyways, the PR is now ready to review. Please only check the recent commit. It will clearly show what the code I have changed. Thanks :)

armanbilge reviewed (commented):
Thanks for your work on this!

May I suggest an alternative approach? I think we should make `ScalQueue` no longer generic. Internally, it can be backed by queues of `AnyRef`. Then, we can adjust its API to be:
```scala
def offer(a: Runnable, random: ThreadLocalRandom): Unit
def offer(a: Array[Runnable], random: ThreadLocalRandom): Unit
```
Then, we can keep the counters inside of `ScalQueue`, and increment the appropriate counters in the specific `offer` methods.

`poll` will have to return `AnyRef` and use `isInstanceOf` checks to decrement the appropriate counter.

Atharva-Kanherkar commented:
> Thanks for your work on this!
> 
> May I suggest an alternative approach? I think we should make `ScalQueue` no longer generic. Internally, it can be backed by queues of `AnyRef`. Then, we can adjust its API to be:
> 
> ```scala
> def offer(a: Runnable, random: ThreadLocalRandom): Unit
> def offer(a: Array[Runnable], random: ThreadLocalRandom): Unit
> ```
> 
> Then, we can keep the counters inside of `ScalQueue`, and increment the appropriate counters in the specific `offer` methods.
> 
> `poll` will have to return `AnyRef` and use `isInstanceOf` checks to decrement the appropriate counter.

For sure. Give  me a day, i will look into it!

Atharva-Kanherkar committed:
Added the requested changes

Atharva-Kanherkar committed:
Added the changes requested

Atharva-Kanherkar commented:
Hey @armanbilge :) I have added the requested changes. Please review! If there is anything to fix again, please let me know!

armanbilge commented:
Thanks for making those changes! It looks like the code doesn't compile currently.

Atharva-Kanherkar commented:
> Thanks for making those changes! It looks like the code doesn't compile currently.

Yes! I am sorry, I actually slept after pushing the changes and forgot fixing the issues. I will be doing it right away! ~Atharva

Atharva-Kanherkar commented:
@armanbilge Actually there were so many references and inter dependencies of how scalqueue was generic, that is why this happened. It seems like there is a lot of code which is written kept in mind that it is generic. However, Ill fix it soon:)

Atharva-Kanherkar commented:
Update : 
![image](https://github.com/user-attachments/assets/d96930f1-8d55-4562-956a-7efaa5a54d2d)
I guess the code works as we wanted it to work! There are still warnings and formatting issues, which can be fixed promptly, but the functionality is working fine now.

Atharva-Kanherkar committed:
Fixed compile errors

Atharva-Kanherkar committed:
Fixed warnings and formatting issues

Atharva-Kanherkar committed:
Fixed CI errors

Atharva-Kanherkar committed:
Fixed CI errors

Atharva-Kanherkar committed:
fixed more CI errrors

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -80,7 +87,15 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
     queues(idx).offer(a)
-    ()
+
+    // Track metrics - if it's a runnable but not an array, it's a singleton task
+    if (!a.isInstanceOf[Array[?]]) {
```
Instead of doing the `isInstanceOf` check, we could introduce a new method `offerBatch` that accepts an `Array[A]`.

Atharva-Kanherkar replied:
Could you please elaborate?

Atharva-Kanherkar replied:
Does this mean, that 
1. Keep the current `offer` method for singleton tasks only
2.Add a new `offerBatch` method specifically for batches
3.Remove the type checking from the existing methods

armanbilge replied:
Yes. So then we would have `offer`, `offerBatch`, and `offerAll`.

Atharva-Kanherkar replied:
okay

Atharva-Kanherkar committed:
Fixed more CI errors

Atharva-Kanherkar committed:
Move external queue metrics tracking to ScalQueue

Atharva-Kanherkar committed:
fixed ci errors

Atharva-Kanherkar committed:
Fixed different CI errors again

Atharva-Kanherkar committed:
Fixed CI errors

Atharva-Kanherkar committed:
fixed other ci errors

Atharva-Kanherkar committed:
Fixed CI errors

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -114,6 +143,11 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
       i += 1
     }
+
+    // Track as batch submissions
+    batchesSubmittedCount.incrementAndGet();
+    batchesPresentCount.incrementAndGet();
```
Actually, I think this should be counted as singletons. Or, we should rename this method to `offerAllBatches` and take an `Array[Array[A]]` as its argument.

Atharva-Kanherkar replied:
Taking your suggestion, I have implemented them as singletons, as : 
 

```// Track as individual submissions (len singletons)
 singletonsSubmittedCount.addAndGet(len)
singletonsPresentCount.addAndGet(len)
    ()
```

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -303,6 +303,13 @@ private[effect] final class WorkerThread[P <: AnyRef](
+   * @return
+   *   reference to the owning WorkStealingThreadPool
+   */
+  private[unsafe] def getPool(): WorkStealingThreadPool[P] = pool
```
Do we still need this method?

Atharva-Kanherkar replied:
Another Good catch! The `getPool() `method isn't being used anywhere in the codebase. I'll remove it in the next commit.

armanbilge reviewed (commented):
Let's change this to a `ScalQueue[Runnable]`, I think that will help us keep track of singletons vs batches.

https://github.com/typelevel/cats-effect/blob/7102a235fd8beeeb8b13131552706c4127b0fb7a/core/jvm/src/main/scala/cats/effect/unsafe/WorkStealingThreadPool.scala#L125-L126

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -220,5 +276,42 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
+   * @return
+   *   a new Scal queue instance
+   */
+  def apply[A <: AnyRef](threadCount: Int): ScalQueue[A] = new ScalQueue(threadCount)
```
Why did we add this method?

Atharva-Kanherkar replied:
The apply method was added to follow the standard Scala idiom for object creation through companion objects, It's being used in `WorkStealingThreadPool.scala` line 241 to create the `externalQueue`.

armanbilge replied:
I see! In Cats Effect's private internal code, we often don't follow Scala idioms and instead write code optimized for performance. In this case, we prefer using `new` instead of an `apply` method.

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkStealingThreadPool.scala:
```diff
@@ -529,6 +527,43 @@ private[effect] final class WorkStealingThreadPool[P <: AnyRef](
+   * @return
+   *   true if the batch was successfully offered
+   */
+  private[unsafe] def offerBatchToExternalQueue(
+      batch: Array[Runnable],
+      random: ThreadLocalRandom): Boolean = {
+    externalQueue.offerBatch(batch, random)
+    true // Assume success
+  }
+
+  /**
+   * Offers multiple batches of runnables to the external queue and updates batch metrics.
+   *
+   * @param batches
+   *   the batches of runnables to be offered to the external queue
+   * @param random
+   *   a reference to an uncontended source of randomness
+   * @return
+   *   true if the batches were successfully offered
+   */
+  private[unsafe] def offerAllBatchesToExternalQueue(
```
Are these methods still used?

Atharva-Kanherkar replied:
Good catch! They are, redundant now. I used grep to find out if they were being called anywhere else and no, they arent. I think we can safely remove these.

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkStealingThreadPool.scala:
```diff
@@ -844,6 +879,37 @@ private[effect] final class WorkStealingThreadPool[P <: AnyRef](
     sum
   }
+
+  /**
+   * Returns the total number of singleton tasks submitted to the external queue.
+   */
+  private[unsafe] def getSingletonsSubmittedCount(): Long =
+    externalQueue.getSingletonsSubmittedCount()
+
+  /**
+   * Returns the total number of batch tasks submitted to the external queue.
+   */
+  private[unsafe] def getBatchesSubmittedCount(): Long =
+    externalQueue.getBatchesSubmittedCount()
+
+  /**
+   * Returns the number of singleton tasks currently in the external queue.
+   */
+  private[unsafe] def getSingletonsPresentCount(): Long =
+    externalQueue.getSingletonsPresentCount()
+
+  /**
+   * Returns the number of batch tasks currently in the external queue.
+   */
+  private[unsafe] def getBatchesPresentCount(): Long = externalQueue.getBatchesPresentCount()
+
+  private[unsafe] def logQueueMetrics(): Unit = {
+    println(s"[Thread Pool ${id}] Queue Metrics:")
+    println(s"  Singletons submitted: ${getSingletonsSubmittedCount()}")
+    println(s"  Singletons present: ${getSingletonsPresentCount()}")
+    println(s"  Batches submitted: ${getBatchesSubmittedCount()}")
+    println(s"  Batches present: ${getBatchesPresentCount()}")
+  }
```
Instead of exposing these on the WSTP, let's make a new `ExternalQueueMetrics` interface in this file:

https://github.com/typelevel/cats-effect/blob/series/3.x/core/jvm/src/main/scala/cats/effect/unsafe/metrics/WorkStealingThreadPoolMetrics.scala

armanbilge commented on tests/jvm/src/main/scala/cats/effect/unsafe/WorkStealingPoolMetricsDemo.scala:
```diff
@@ -0,0 +1,142 @@
+ * Demo program to verify that the WorkStealingThreadPool metrics for singletons and batches are
+ * working correctly.
+ */
+object WorkStealingPoolMetricsDemo {
```
This is cool! Do you think we can refactor this as unit tests?

Atharva-Kanherkar replied:
Yes! I wanted it to, but I ran into Null Pointer Exceptions and could not fix it. I just made it to make sure my code worked. I think with some help, I can for sure make it as a test.

Atharva-Kanherkar commented:
> Let's change this to a `ScalQueue[Runnable]`, I think that will help us keep track of singletons vs batches.
> 
> https://github.com/typelevel/cats-effect/blob/7102a235fd8beeeb8b13131552706c4127b0fb7a/core/jvm/src/main/scala/cats/effect/unsafe/WorkStealingThreadPool.scala#L125-L126

With this change, we'll also need to update related parts of the code:


In the `liveTraces() `method (line ~541):
In the `stealFromOtherWorkerThread` method (line ~317):
Is it okay if I change these methods too? For the record, I have changed it to Runnable.

armanbilge commented:
> Is it okay if I change these methods too? For the record, I have changed it to Runnable.

Yes, that makes sense!

Atharva-Kanherkar commented:
Hi @armanbilge,

I've implemented the changes you suggested:

1. Added a new `ExternalQueueMetrics` trait in the metrics package
2. Added the `externalQueue` property to `WorkStealingThreadPoolMetrics`
3. Removed the metrics methods from `WorkStealingThreadPool`
4. Updated `offerAll` in `ScalQueue` to track as individual submissions
5. Removed the unused methods (`getPool()`, `offerBatchToExternalQueue`, etc.)

However, I'm running into some type mismatch errors:

1. When trying to change `externalQueue` from `ScalQueue[AnyRef]` to `ScalQueue[Runnable]`:
`[error] found: cats.effect.unsafe.ScalQueue[Runnable] [error] required: cats.effect.unsafe.ScalQueue[AnyRef]`
2. In `liveTraces()` method when processing elements from the queue:
`[error] found: AnyRef [error] required: Runnable`

I see two possible solutions:
1. Keep `externalQueue` as `ScalQueue[AnyRef]` and continue to do type checking
2. Make `ScalQueue` covariant in its type parameter (define as `ScalQueue[+A <: AnyRef]`)

Which approach would you prefer? The covariance change would be larger but could simplify the code overall.

Thank you!

armanbilge commented:
I think we should use `ScalQueue[Runnable]`, but we don't need to make any covariance changes. In all the places where there is an type mismatch, that means we should change the type :) if that still doesn't make sense, push your code up and I can help you work through the errors.

Atharva-Kanherkar commented:
> I think we should use `ScalQueue[Runnable]`, but we don't need to make any covariance changes. In all the places where there is an type mismatch, that means we should change the type :) if that still doesn't make sense, push your code up and I can help you work through the errors.

I'll push my code up because this got a little, a little only;) confusing

Atharva-Kanherkar committed:
Added requested changes

armanbilge commented:
@Atharva-Kanherkar I think you are getting compile errors, because you didn't change it in `WorkerThread`.

https://github.com/typelevel/cats-effect/blob/1f084bf6aaafd179fe64b4d004521c91e81e2eb4/core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala#L54

You may need to change it in other places as well.

Atharva-Kanherkar commented:
> @Atharva-Kanherkar I think you are getting compile errors, because you didn't change it in `WorkerThread`.
> 
> https://github.com/typelevel/cats-effect/blob/1f084bf6aaafd179fe64b4d004521c91e81e2eb4/core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala#L54
> 
> You may need to change it in other places as well.

Yes, but this would require changing many other parts of the code, including method signatures and usage patterns, and could introduce more bugs.
Are you sure we can proceed with this?
 To maintain compatibility with the existing codebase, which uses `ScalQueue[AnyRef]` consistently throughout. we  can't change just one part to use `ScalQueue[Runnable] `without changing all the related parts, which would be a much larger and riskier change.
I dont have a problem doing that, Just hoping to get a green flag on this one :)

armanbilge commented:
Yes, please go ahead and make the change 🙂

Atharva-Kanherkar committed:
Added the requested changes, without the test

Atharva-Kanherkar committed:
Added headers

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -362,35 +362,14 @@ private[effect] final class WorkerThread[P <: AnyRef](
         // Check the external queue after a failed dequeue from the local
         // queue (due to the local queue being empty).
         val element = external.poll(rnd)
-        if (element.isInstanceOf[Array[Runnable]]) {
-          val batch = element.asInstanceOf[Array[Runnable]]
-          // The dequeued element was a batch of fibers. Enqueue the whole
-          // batch on the local queue and execute the first fiber.
-          // It is safe to directly enqueue the whole batch because we know
-          // that in this state of the worker thread state machine, the
-          // local queue is empty.
-          val fiber = queue.enqueueBatch(batch, self)
-          // Many fibers have been exchanged between the external and the
-          // local queue. Notify other worker threads.
-          pool.notifyParked(rnd)
-          try fiber.run()
-          catch {
-            case t if UnsafeNonFatal(t) => pool.reportFailure(t)
-            case t: Throwable => IOFiber.onFatalFailure(t)
-          }
-
-          // Transition to executing fibers from the local queue.
-          return
-        } else if (element.isInstanceOf[Runnable]) {
-          val fiber = element.asInstanceOf[Runnable]
```
Why is all of this code deleted?

Atharva-Kanherkar replied:
I am so sorry! You concern is totally  valid -  I've removed the batch handling code in the `lookForWork() `method. Even when using `ScalQueue[Runnable]`, the queue can still contain `Array[Runnable] `objects that need to be handled specially. I was trying to do eradicate some instances, and this might have happened, Ill push it back

Atharva-Kanherkar commented:
hey @armanbilge , I am getting some fruitless warnings here : 
![image](https://github.com/user-attachments/assets/b921115e-4b51-44e7-8f76-6d69b5863217)
With your permission, can I use @unchecked here? Something like this :  
```
val element = external.poll(rnd)
if ((element: AnyRef).isInstanceOf[Array[Runnable @unchecked]]) {
  val batch = element.asInstanceOf[Array[Runnable]]
  // ...rest of code...
}
```

Atharva-Kanherkar committed:
Removed warnings

Atharva-Kanherkar committed:
Added the requested changes

Atharva-Kanherkar commented:
@armanbilge I've completed the changes to make ScalQueue non-generic with a type-safe API that distinguishes between singleton and batch tasks:

1. Removed the type parameter from ScalQueue, making it non-generic
   - Changed from `ScalQueue[A <: AnyRef] `to just `ScalQueue`

2. Updated internal implementation to use `AnyRef` for heterogeneous storage:
   - `private[this] val queues: Array[ConcurrentLinkedQueue[AnyRef]]
`
3. Created a type-safe public API with specialized methods:
   - `offer(a: Runnable, random: ThreadLocalRandom): Unit `- for singleton tasks
   - `offerBatch(batch: Array[Runnable], random: ThreadLocalRandom): Unit` - for batch tasks
   - `offerAll(as: Array[Runnable], random: ThreadLocalRandom): Unit` - for multiple singletons

4. Added metrics tracking in appropriate methods:
   - In `offer(): `increment `singletonsSubmittedCount` & `singletonsPresentCount`
   - In `offerBatch():` increment `batchesSubmittedCount` & `batchesPresentCount`
   - In `poll():` decrement the appropriate counter based on element type

5. Updated all references to `ScalQueue` throughout the codebase:
   - `WorkStealingThreadPool.scala`
   - `WorkerThread.scala`
   - `LocalQueue.scala`
   - `ScalQueueBenchmark.scala`
   - `WorkStealingThreadPoolMetrics.scala`

6. Added proper type checking and casting in code that uses `ScalQueue`:
   - `element.asInstanceOf[Runnable] `for singleton tasks
   - `element.asInstanceOf[Array[Runnable]] `for batch tasks

 This keeps all the performance-focused design choices intact while improving type safety at the API level. 🚀 Let me know if you have any thoughts! If I have forgotten something or something is wrong, please let me know!

Atharva-Kanherkar committed:
Trying to fix the CI

Atharva-Kanherkar committed:
Formatted

Atharva-Kanherkar committed:
Fixed tests

Atharva-Kanherkar commented:
As I am running into the issues, i'd like to document them for future ease.

The errors specifically mention two methods:

`offer(java.lang.Object, ThreadLocalRandom)` changed to `offer(java.lang.Runnable, ThreadLocalRandom)
``offerAll(Array[java.lang.Object], ThreadLocalRandom) `changed to `offerAll(Array[java.lang.Runnable], ThreadLocalRandom)
`
## Why I think, these changes Are important : 
- Type Safety: The original design used `ScalQueue[AnyRef] `which wasn't type-safe. By making the API methods explicitly accept `Runnable` and `Array[Runnable] `, we've improved type safety.

- Clear API Distinction: The changes make it explicit what types of elements can be enqueued:

`offer(Runnable, ...) `- for individual tasks
`offerBatch(Array[Runnable], ...) `- for batch tasks

- Performance Focus: As @armanbilge said, _In Cats Effect's internal runtime, we often make 'non-idiomatic' choices to optimize for performance._ The new design maintains this by:

1. Using `AnyRef` internally for heterogeneous storage (avoiding wrappers)
2. Keeping direct access to elements without additional indirection
3. Avoiding `Either` or wrapper types that would require extra allocations

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/LocalQueue.scala:
```diff
@@ -281,9 +281,10 @@ private final class LocalQueue extends LocalQueuePadding {
-        external.offerAll(batches, random)
-        // Loop again for a chance to insert the original fiber to be enqueued
-        // on the local queue.
+        //  (loop through each batch)
+        for (batch <- batches) {
+          external.offerBatch(batch, random)
+        }
```
Instead of making this change here, let's introduce a method `offerAllBatches` on `ScalQueue`.

Atharva-Kanherkar replied:
okay~!

Atharva-Kanherkar replied:
This is done! Ill push the code soon!

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -104,16 +133,22 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
       val fiber = as(i)
       val idx = random.nextInt(nq)
-      queues(idx).offer(fiber)
+      queues(idx).offer(fiber.asInstanceOf[AnyRef]) // Cast to AnyRef
```
Why do we need a cast here?

Atharva-Kanherkar replied:
The cast to `AnyRef` is needed because we've changed the internal implementation of the queues to use `AnyRef`:

`private[this] val queues: Array[ConcurrentLinkedQueue[AnyRef]]`

Even though we've updated the API to accept properly typed `Runnable` elements, we need to cast them to `AnyRef` when storing them in the underlying `ConcurrentLinkedQueue`. I was getting compiler errors, without this cast, so I did it. Is it causing any optimization issues?

armanbilge replied:
Hmm, are you certain? A `Runnable` is already any `AnyRef` so it should not need to be casted.

Atharva-Kanherkar replied:
Yes I am very sure. I can reverse engineer this and show you the screenshot of the errors if you'd like.

armanbilge replied:
I think you are going to have to share a screenshot, I can't reproduce your compile error 😕

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -125,19 +160,30 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
     val from = random.nextInt(nq)
     var i = 0
-    var a = null.asInstanceOf[A]
+    var element = null.asInstanceOf[AnyRef]
```
```suggestion
    var element: AnyRef = null
```

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -125,19 +160,30 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
+    if (element != null) {
+      // We still need to check the type here since we don't know whether we're
+      // dequeuing a singleton or a batch
+      if (element.isInstanceOf[Array[?]]) {
```
```suggestion
      if (element.isInstanceOf[Array[Runnable]]) {
```

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -165,6 +211,17 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
+    if (done) {
+      // We still need to check the type here since we don't know whether we're
+      // removing a singleton or a batch
+      if (a.isInstanceOf[Array[?]]) {
```
```suggestion
      if (a.isInstanceOf[Array[Runnable]]) {
```

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -173,8 +230,12 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
    * @return
    *   a set of the currently enqueued elements
    */
-  def snapshot(): Set[AnyRef] =
-    queues.flatMap(_.toArray).toSet
+  def snapshot(): Array[Runnable] = {
+    val elements = queues.flatMap(_.toArray)
+
+    // Filter out elements that aren't Runnable
+    elements.collect { case r: Runnable => r }.toArray
+  }
```
I don't think we need to change this method.

Atharva-Kanherkar replied:
This method caused a test to fail. I think I can show you the screen shot for this too, and we can try fixing it!

Atharva-Kanherkar replied:
#  `cats.effect.unsafe.FiberMonitorSuite`

```plaintext
[error] Failed: Total 2918, Failed 1, Errors 0, Passed 2917, Ignored 1
[error] Failed tests:
[error] cats.effect.unsafe.FiberMonitorSuite
[error] (testsJVM / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 371 s (0:06:11.0), completed Mar 12, 2025, 9:58:56 AM

[error] ==> X cats.effect.unsafe.FiberMonitorSuite.show only active fibers in a live snapshot 0.107s
java.lang.ClassCastException: class scala.collection.immutable.Set$EmptySet$ 
    cannot be cast to class [Ljava.lang.Runnable; 
    (scala.collection.immutable.Set$EmptySet$ is in unnamed module of loader 'app'; 
    [Ljava.lang.Runnable; is in module java.base of loader 'bootstrap')
[error]  at apply @ munit.internal.console.StackTraces$.dropOutside(StackTraces.scala:12)
[error]  at $anonfun$realWithRuntime$1 @ munit.internal.console.StackTraces$.dropOutside(StackTraces.scala:12)

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -220,5 +281,42 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
+   * @return
+   *   a new Scal queue instance
+   */
+  def apply(threadCount: Int): ScalQueue = new ScalQueue(threadCount)
```
We don't need to add this, `new ScalQueue` is perfectly fine for internal use and avoids the indirection.

Atharva-Kanherkar replied:
Okay! for sure :)

Atharva-Kanherkar replied:
This is done. Sorry I forgot about this, knowing it already.

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkStealingThreadPool.scala:
```diff
@@ -234,18 +232,13 @@ private[effect] final class WorkStealingThreadPool[P <: AnyRef](
       if (isStackTracing) {
-        destWorker.active = fiber
+        destWorker.active = element.asInstanceOf[Runnable]
         parkedSignals(dest).lazySet(false)
       }
-
-      fiber
+      element.asInstanceOf[Runnable]
```
I think you want to revert these changes.

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkStealingThreadPool.scala:
```diff
@@ -545,13 +539,13 @@ private[effect] final class WorkStealingThreadPool[P <: AnyRef](
-        case batch: Array[Runnable] =>
-          batch.flatMap(r => captureTrace(r)).toMap[Runnable, Trace]
-        case r: Runnable => captureTrace(r).toMap[Runnable, Trace]
-        case _ => Map.empty[Runnable, Trace]
-      }
+      .flatMap(r =>
+        captureTrace(r) match {
+          case Some((_, trace)) => Some((r, trace))
+          case None => None
+        })
```
Same, I think we can revert these changes.

Atharva-Kanherkar replied:
Okay! I agree this was not necessary, This must be before we talked about the cats effect in detail, and I did not remember to refactor this to the original. I have done it now!

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -362,7 +362,7 @@ private[effect] final class WorkerThread[P <: AnyRef](
         // queue (due to the local queue being empty).
         val element = external.poll(rnd)
-        if (element.isInstanceOf[Array[Runnable]]) {
+        if ((element: AnyRef).isInstanceOf[Array[Runnable]]) {
```
Why do we need this change?

Atharva-Kanherkar replied:
We do not! I am reverting the change!

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -381,16 +381,15 @@ private[effect] final class WorkerThread[P <: AnyRef](
+        } else if (element != null) {
+          // Existing code for handling individual Runnable
           if (isStackTracing) {
-            _active = fiber
+            _active = element.asInstanceOf[Runnable]
             parked.lazySet(false)
           }
 
-          // The dequeued element is a single fiber. Execute it immediately.
-          try fiber.run()
+          // The dequeued element is a Runnable (fiber). Execute it immediately.
+          try element.asInstanceOf[Runnable].run()
```
I think we can revert these changes.

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -476,18 +475,14 @@ private[effect] final class WorkerThread[P <: AnyRef](
 
       while (!done.get()) {
-        // Check the external queue after being unparked
         val element = external.poll(rnd)
-        if (element.isInstanceOf[Array[Runnable]]) {
+        if ((element: AnyRef).isInstanceOf[Array[Runnable]]) {
           val batch = element.asInstanceOf[Array[Runnable]]
           // Announce that the current thread is no longer looking for work.
           pool.transitionWorkerFromSearching(rnd)
 
           // The dequeued element was a batch of fibers. Enqueue the whole
           // batch on the local queue and execute the first fiber.
-          // It is safe to directly enqueue the whole batch because we know
-          // that in this state of the worker thread state machine, the
-          // local queue is empty.
```
I don't think we need these changes.

Atharva-Kanherkar replied:
Yes. We do not. I am so sorry. This had to be to fix some of the compile errors. But when those got fixed (internally) i pushed this too. I have reverted it.

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -500,19 +495,18 @@ private[effect] final class WorkerThread[P <: AnyRef](
-
           if (isStackTracing) {
-            _active = fiber
+            _active = element.asInstanceOf[Runnable]
```
Revert changes?

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -500,19 +495,18 @@ private[effect] final class WorkerThread[P <: AnyRef](
-          // The dequeued element is a single fiber. Execute it immediately.
-          try fiber.run()
+          // The dequeued element is a Runnable (fiber). Execute it immediately.
+          try element.asInstanceOf[Runnable].run()
```
Revert changes?

Atharva-Kanherkar replied:
Do you want me to do something like : 

```scala
val fiber = element.asInstanceOf[Runnable]
// I need to Keep this line:
try fiber.run()
```

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -794,9 +788,8 @@ private[effect] final class WorkerThread[P <: AnyRef](
-          // Obtain a fiber or batch of fibers from the external queue.
           val element = external.poll(rnd)
-          if (element.isInstanceOf[Array[Runnable]]) {
+          if ((element: AnyRef).isInstanceOf[Array[Runnable]]) {
```
Revert changes?

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -815,16 +808,14 @@ private[effect] final class WorkerThread[P <: AnyRef](
-            // The dequeued element is a single fiber. Execute it immediately.
-            try fiber.run()
+            // The dequeued element is a Runnable (fiber). Execute it immediately.
+            try element.asInstanceOf[Runnable].run()
```
Revert changes?

Atharva-Kanherkar replied:
These (and all above) I changed because they were causing compiler errors. I can revert them and tell you what the errors were. I am very sorry if this thing has went wrong. Is it okay if i post the errors on discord?

armanbilge replied:
@Atharva-Kanherkar sure, that's fine.

I know it may seem surprising, but we shouldn't need _that_ many changes at all 🙂 
1. removing the type parameter from `ScalQueue` anywhere it's declared
2. in a few places, you may need to replace `offer` with `offerBatch` (or `offerAll` with `offerAllBatch`)
3. anything involving `poll` doesn't need to be changed, because we didn't change that API

Atharva-Kanherkar replied:
Okay! Fir sure, I might have gone a little too far with the code change. Ill fix these asap.

Atharva-Kanherkar committed:
Update core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

Atharva-Kanherkar committed:
Update core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

Atharva-Kanherkar committed:
Update core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

Atharva-Kanherkar committed:
Requested changes

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/LocalQueue.scala:
```diff
@@ -281,9 +281,7 @@ private final class LocalQueue extends LocalQueuePadding {
-        external.offerAll(batches, random)
-        // Loop again for a chance to insert the original fiber to be enqueued
-        // on the local queue.
+        external.offerAllBatches(batches, random)
```
The code comment is still relevant, so we shouldn't delete it.

Atharva-Kanherkar replied:
Okay!!

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -70,16 +77,38 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
-  def offer(a: A, random: ThreadLocalRandom): Unit = {
+  def offer(a: Runnable, random: ThreadLocalRandom): Unit = {
+    val idx = random.nextInt(numQueues)
+    queues(idx).offer(a.asInstanceOf[AnyRef])
```
```suggestion
    queues(idx).offer(a)
```

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -70,16 +77,38 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
+  def offerBatch(batch: Array[Runnable], random: ThreadLocalRandom): Unit = {
     val idx = random.nextInt(numQueues)
-    queues(idx).offer(a)
+    queues(idx).offer(batch.asInstanceOf[AnyRef])
```
```suggestion
    queues(idx).offer(batch)
```

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -220,5 +304,42 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
+  def getBatchesPresentCount(): Long = batchesPresentCount.get()
+}
+
+object ScalQueue {
+
+  /**
+   * Creates a new Scal queue.
+   *
+   * @param threadCount
+   *   the number of threads to load balance between
+   * @return
+   *   a new Scal queue instance
+   */
+
```
This can be cleaned up :)

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -476,7 +476,6 @@ private[effect] final class WorkerThread[P <: AnyRef](
       }
 
       while (!done.get()) {
-        // Check the external queue after being unparked
```
This comment is still relevant and should not be deleted.

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -485,9 +484,6 @@ private[effect] final class WorkerThread[P <: AnyRef](
           // batch on the local queue and execute the first fiber.
-          // It is safe to directly enqueue the whole batch because we know
-          // that in this state of the worker thread state machine, the
-          // local queue is empty.
           val fiber = queue.enqueueBatch(batch, self)
```
This comment is still relevant and should not be deleted.

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -794,7 +790,6 @@ private[effect] final class WorkerThread[P <: AnyRef](
           // update the current time
           now = System.nanoTime()
         } else {
-          // Obtain a fiber or batch of fibers from the external queue.
```
This comment is still relevant and should not be deleted.

Atharva-Kanherkar replied:
Okay, done!

Atharva-Kanherkar committed:
Update core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

Atharva-Kanherkar committed:
Update core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

Atharva-Kanherkar committed:
Added requested changes

Atharva-Kanherkar committed:
Implement striped metrics counters in ScalQueue

Atharva-Kanherkar committed:
Fixed warnings

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/metrics/WorkStealingThreadPoolMetrics.scala:
```diff
@@ -85,12 +85,54 @@ sealed trait WorkStealingThreadPoolMetrics {
+  /**
+   * Returns the total number of singleton tasks submitted to the queue.
+   */
+  def singletonsSubmittedCount(): Long
+
+  /**
+   * Returns the number of singleton tasks currently in the queue.
+   */
+  def singletonsPresentCount(): Long
+
+  /**
+   * Returns the total number of batch tasks submitted to the queue.
+   */
+  def batchesSubmittedCount(): Long
+
+  /**
+   * Returns the number of batch tasks currently in the queue.
+   */
+  def batchesPresentCount(): Long
```
This matches our existing naming convention. Also, I realized we should add two more metrics: `fiberCount` and `totalFiberCount`, that counts singletons and the number of fibers in the batches.

```suggestion
  def totalSingletonCount(): Long

  /**
   * Returns the number of singleton tasks currently in the queue.
   */
  def singletonCount(): Long

  /**
   * Returns the total number of batch tasks submitted to the queue.
   */
  def totalBatchCount(): Long

  /**
   * Returns the number of batch tasks currently in the queue.
   */
  def batchCount(): Long
```

Atharva-Kanherkar replied:
Okay, for sure!

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/metrics/WorkStealingThreadPoolMetrics.scala:
```diff
@@ -85,12 +85,54 @@ sealed trait WorkStealingThreadPoolMetrics {
+  def batchesPresentCount(): Long
+}
+
+object ExternalQueueMetrics {
+
+  private[metrics] def apply(queue: ScalQueue): ExternalQueueMetrics =
```
You can just move this inside of the `object WorkStealingThreadPoolMetrics`.

Atharva-Kanherkar replied:
Okay, moved!!!

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -55,12 +56,22 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
 
+  // Metrics counters for tracking external queue submissions
+  private[this] val singletonsSubmittedCounts: Array[AtomicLong] =
+    Array.fill(numQueues)(new AtomicLong(0))
```
Instead of `Array.fill`, let's initialize the arrays with a `while` loop like how `queues` is below.

Atharva-Kanherkar replied:
Okay. Done!

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -125,19 +193,34 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
-      a = queues(idx).poll()
+      element = queues(idx).poll()
+      if (element ne null) {
+        pollIdx = idx // Remember which queue we got the element from
```
Instead of passing the index, why don't we just do the decrement here?

Atharva-Kanherkar replied:
Okay, I will implement the change!

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -156,15 +239,30 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
     while (!done && i < nq) {
       done = queues(i).remove(a)
+      if (done) {
+        removeIdx = i // Remember which queue the element was removed from
```
Same comment, let's just decrement the counter here?

Atharva-Kanherkar replied:
Okay, done!

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala:
```diff
@@ -488,6 +488,7 @@ private[effect] final class WorkerThread[P <: AnyRef](
           // It is safe to directly enqueue the whole batch because we know
           // that in this state of the worker thread state machine, the
           // local queue is empty.
+
```
Spurious change

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/LocalQueue.scala:
```diff
@@ -281,9 +281,10 @@ private final class LocalQueue extends LocalQueuePadding {
+        external.offerAllBatches(batches, random)
         // Loop again for a chance to insert the original fiber to be enqueued
         // on the local queue.
+
```
Spurious change

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkStealingThreadPool.scala:
```diff
@@ -249,6 +247,7 @@ private[effect] final class WorkStealingThreadPool[P <: AnyRef](
     } else {
       null
     }
+
```
Spurious change

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/WorkStealingThreadPool.scala:
```diff
@@ -529,6 +528,7 @@ private[effect] final class WorkStealingThreadPool[P <: AnyRef](
     externalQueue.offer(fiber, random)
     notifyParked(random)
     ()
+
```
Spurious change

armanbilge reviewed (commented):
Looking great, now just polishing some stuff!

Atharva-Kanherkar committed:
Update core/jvm/src/main/scala/cats/effect/unsafe/metrics/WorkStealingThreadPoolMetrics.scala

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

Atharva-Kanherkar committed:
Added requested changes

Atharva-Kanherkar committed:
Fixed warnings

Atharva-Kanherkar committed:
FIXED WARNINGS

Atharva-Kanherkar commented:
@armanbilge Does it seem okay now, The CI Is green, tell if you need any more changes!:)

Atharva-Kanherkar committed:
Removed un necessary comments

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/metrics/WorkStealingThreadPoolMetrics.scala:
```diff
@@ -244,6 +286,17 @@ sealed trait TimerHeapMetrics {
 }
 
 object WorkStealingThreadPoolMetrics {
+  private object ExternalQueueMetrics {
```
Thanks for moving it here! What I meant to suggest was we make it a private method, instead of an object. You can follow the existing pattern used for all the other metrics in this class (see the code below).

Atharva-Kanherkar replied:
Okay!

Atharva-Kanherkar replied:
This is done! I have - 
 
- Removed this implementation : 
```scala
private object ExternalQueueMetrics {
  private[metrics] def apply(queue: ScalQueue): ExternalQueueMetrics =
    new ExternalQueueMetrics {
      def totalSingletonCount(): Long = queue.getSingletonsSubmittedCount()
      def singletonCount(): Long = queue.getSingletonsPresentCount()
      def totalBatchCount(): Long = queue.getBatchesSubmittedCount()
      def batchCount(): Long = queue.getBatchesPresentCount()
      def totalFiberCount(): Long = queue.getTotalFiberSubmittedCount()
      def fiberCount(): Long = queue.getFiberPresentCount()
    }
}
```
- Changed it to : 
```scala
private def externalQueueMetrics(queue: ScalQueue): ExternalQueueMetrics =
  new ExternalQueueMetrics {
    def totalSingletonCount(): Long = queue.getSingletonsSubmittedCount()
    def singletonCount(): Long = queue.getSingletonsPresentCount()
    def totalBatchCount(): Long = queue.getBatchesSubmittedCount()
    def batchCount(): Long = queue.getBatchesPresentCount()
    def totalFiberCount(): Long = queue.getTotalFiberSubmittedCount()
    def fiberCount(): Long = queue.getFiberPresentCount()
  }
```
This style is consistent with the other code in the same file.

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/metrics/WorkStealingThreadPoolMetrics.scala:
```diff
@@ -244,6 +286,17 @@ sealed trait TimerHeapMetrics {
+  private object ExternalQueueMetrics {
+    private[metrics] def apply(queue: ScalQueue): ExternalQueueMetrics =
+      new ExternalQueueMetrics {
+        def totalSingletonCount(): Long = queue.getSingletonsSubmittedCount()
+        def singletonCount(): Long = queue.getSingletonsPresentCount()
+        def totalBatchCount(): Long = queue.getBatchesSubmittedCount()
+        def batchCount(): Long = queue.getBatchesPresentCount()
+        def totalFiberCount(): Long = queue.getTotalFiberSubmittedCount()
+        def fiberCount(): Long = queue.getFiberPresentCount()
```
Can we adjust the method names in `ScalQueue` so they correspond better with the names of the metrics?

Atharva-Kanherkar replied:
Okay, done!

Atharva-Kanherkar committed:
Adding requested changes

Atharva-Kanherkar committed:
Added test

Atharva-Kanherkar committed:
Removed not important file

Atharva-Kanherkar committed:
Reformatting

Atharva-Kanherkar committed:
Some changes

Atharva-Kanherkar committed:
Fixed warnings

Atharva-Kanherkar committed:
Added tests

Atharva-Kanherkar committed:
Fixed errors

armanbilge commented on tests/jvm/src/test/scala/cats/effect/unsafe/ScalQueueSuite.scala:
```diff
@@ -0,0 +1,126 @@
+    val random = ThreadLocalRandom.current()
+
+    // Get initial metrics before any operations
+    val initialSingletonCount = queue.getTotalSingletonCount()
+    val initialBatchCount = queue.getTotalBatchCount()
+    val initialFiberCount = queue.getTotalFiberCount()
```
These should all be zero, right?

Atharva-Kanherkar replied:
Yes, This slipped out of my mind. They shud be zero cause the queue has just been created. I'll add assertions to verify that the initial values are indeed zero. Does it sound good?

armanbilge commented on tests/jvm/src/test/scala/cats/effect/unsafe/ScalQueueSuite.scala:
```diff
@@ -0,0 +1,126 @@
+    // Add the batch to the queue
+    queue.offerBatch(batch, random)
+
+    // Verify that the batch count has increased
+    val afterBatchCount = queue.getTotalBatchCount()
+    assert(afterBatchCount > initialBatchCount)
```
It should be exactly one now, right?

Atharva-Kanherkar replied:
Yes, right! I'll update this!

Atharva-Kanherkar replied:
`assertEquals(afterBatchCount, initialBatchCount + 1, "Batch count should increase by exactly 1")`
Is this okay for making it exactly one?

armanbilge replied:
Why not just `assertEquals(afterBatchCount, 1)` because we know `initialBatchCount` is 0?

Atharva-Kanherkar replied:
thats trivial! i dont know why this didnt come in my mind! yes you are right - it can be just that! so sorry!

Atharva-Kanherkar replied:
@armanbilge I have added the change!

Atharva-Kanherkar committed:
ensured code quality

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -55,12 +56,74 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
   private[this] val numQueues: Int = mask + 1
 
+  // Metrics counters for tracking external queue submissions
+  private[this] val singletonsSubmittedCounts: Array[AtomicLong] = {
+    val counts = new Array[AtomicLong](numQueues)
+    var i = 0
+    while (i < numQueues) {
+      counts(i) = new AtomicLong(0)
+      i += 1
+    }
+    counts
+  }
```
Since this code is duplicated, we can extract into a method to allocate and initialize an `Array[AtomicLong]`.

Atharva-Kanherkar replied:
Okay for sure!

Atharva-Kanherkar replied:
I have implementd the change..

armanbilge commented on core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -214,11 +386,96 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
+  /**
+   * Returns the total number of singleton tasks submitted to this queue.
+   */
+  def getTotalSingletonCount(): Long = {
+    var total = 0L
+    var i = 0
+    while (i < numQueues) {
+      total += singletonsSubmittedCounts(i).get()
```
Let's make all the names consistent.
```suggestion
  def getTotalSingletonCount(): Long = {
    var total = 0L
    var i = 0
    while (i < numQueues) {
      total += totalSingletonCounts(i).get()
```

Atharva-Kanherkar replied:
I am sorry for that, I'll fix this.

Atharva-Kanherkar replied:
Hey @armanbilge , I was getting confused with the names. I will update it all, being consistent in the codebase.

armanbilge commented on tests/jvm/src/test/scala/cats/effect/unsafe/ScalQueueSuite.scala:
```diff
@@ -0,0 +1,161 @@
+  /**
+   * Tests that the ScalQueue metrics correctly track singleton and batch submissions.
+   */
+  test("ScalQueue metrics track singleton and batch counts") {
```
May I suggest breaking this large test into several smaller tests?

- offering singleton
- offering batch
- polling singleton
- polling batch

Atharva-Kanherkar committed:
Update core/jvm/src/main/scala/cats/effect/unsafe/ScalQueue.scala

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

Atharva-Kanherkar committed:
Changes

Atharva-Kanherkar committed:
Added requested changes

Atharva-Kanherkar committed:
Changed names

Atharva-Kanherkar committed:
Fixed build.sbt

Atharva-Kanherkar commented:
Hello @armanbilge ! I have added the changes! Please review :)

Atharva-Kanherkar committed:
Merge remote-tracking branch 'origin/series/3.x' into external-queue-tracking

Atharva-Kanherkar commented:
Hey @armanbilge!
Just wanted to follow up on this PR — marked it as ready a couple days ago. No rush at all, but when you get a chance, your review would be super appreciated! 🙌

armanbilge commented on build.sbt:
```diff
@@ -725,11 +725,18 @@ lazy val core = crossProject(JSPlatform, JVMPlatform, NativePlatform)
       // internal API change, makes CpuStarvationMetrics available on all platforms
       ProblemFilters.exclude[MissingClassProblem](
         "cats.effect.metrics.JvmCpuStarvationMetrics$NoOpCpuStarvationMetrics"),
+        // This method was added to WorkStealingThreadPoolMetrics to expose queue metrics
+// for monitoring and debugging purposes. It's an addition to the public API that
+// doesn't break existing functionality, as existing implementations can provide
+// a default implementation returning empty metrics.
+ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.effect.unsafe.metrics.WorkStealingThreadPoolMetrics.externalQueue"),
```
This comment is not quite correct.

> as existing implementations can provide a default implementation returning empty metrics

While this is true, that would require existing implementations to change their code to add that implementation. So in general, this is not sufficient to avoid breaking compatibility.

The reason we know this change is safe, is because `WorkStealingThreadPoolMetrics` is a `sealed trait`. That means it can only be implemented in the file that defines it, which means we control all of the implementations. Because we are confident there are no other implementations out in the wild, we don't have to worry about breaking compatibility with them.

You can add a short, explanatory comment similar to this one:

https://github.com/typelevel/cats-effect/blob/11c705a1e11c72dfe25ba4eaea93e56a62bb2503/build.sbt#L717-L719

Atharva-Kanherkar replied:
Okay! Thank you so much for the explanation! 
Is this alright : 
```scala
// introduced by #4292 , external queue metrics
// WorkStealingThreadPoolMetrics is a sealed trait, so we control all of its implementations.
// This ensures that adding this method does not break compatibility. (this is optional , ill likely remove this)
ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.effect.unsafe.metrics.WorkStealingThreadPoolMetrics.externalQueue"),
```

Atharva-Kanherkar replied:
I have changed this.

armanbilge replied:
Thanks! Did you push the changes? I can't see it for some reason.

Atharva-Kanherkar replied:
Yes! I have pushed the changes now!

armanbilge commented on core/jvm-native/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -55,12 +56,37 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
+  private[this] val singletonCounts: Array[AtomicLong] = createAtomicLongArray(numQueues)
+  private[this] val totalBatchCounts: Array[AtomicLong] = createAtomicLongArray(
+    numQueues
+  ) // Change from batchesSubmittedCounts
```
I don't think we need to keep this as a permanent comment in the codebase, since the old names no longer exist anywhere in the code anyway.

Was the comment intended to help me with the review? If so, I appreciate it, but it's not needed, diffs make it pretty easy to see what's changed :)

Atharva-Kanherkar replied:
I made it so that I could remember what changes I was doing, because a lot of changes were confusing (specially regarding names!) and damn, I forgot to remove this one! Mistake from my side! I should have removed this one! Sorry for this!

armanbilge commented on core/jvm-native/src/main/scala/cats/effect/unsafe/ScalQueue.scala:
```diff
@@ -125,19 +226,38 @@ private[effect] final class ScalQueue[A <: AnyRef](threadCount: Int) {
    *   an element from this Scal queue or `null` if this queue is empty
    */
-  def poll(random: ThreadLocalRandom): A = {
+  // Update the poll method (around line 230)
```
Same here.

armanbilge commented on tests/jvm/src/test/scala/cats/effect/unsafe/ScalQueueSuite.scala:
```diff
@@ -0,0 +1,239 @@
+    assertEquals(initialFiberCount, 0L, "Initial fiber count should be zero")
+
+    // Add a singleton task for later fiber count verification
+    queue.offer(new Runnable { def run(): Unit = () }, random)
```
Shorthand:

```suggestion
    queue.offer(() => (), random)
```

armanbilge commented on tests/jvm/src/test/scala/cats/effect/unsafe/ScalQueueSuite.scala:
```diff
@@ -0,0 +1,239 @@
+      i += 1
+    }
+
+    // Verify we were able to poll at least one task
+    assert(polledCount > 0, "Should have polled at least one task")
```
Is there a situation where we didn't poll exactly 10 times?

Atharva-Kanherkar replied:
You're right - in an ideal scenario, we would poll exactly 10 times successfully. However, I used the weaker assertion because the polling behavior can be affected by the random distribution across stripes?
Would you prefer I change it to a stronger assertion like `assertEquals(polledCount, 10)` or perhaps something in between like `assert(polledCount >= 6) ` to ensure we're polling a majority of the expected items?

armanbilge replied:
> because the polling behavior can be affected by the random distribution across stripes?

Hmm, can you explain in more detail? I think it should always return exactly 10 items (assuming there are at least 10 items in the queue). Even though it's randomly striped, poll iterates across the stripes until it finds an element (or there are no more stripes left).

Atharva-Kanherkar replied:
You are actually right. My reasoning wasn't exactly correct/Making sense. I've updated the assertion to `assertEquals(polledCount, 10, "Should have polled exactly 10 tasks")` to properly verify this behavior. Thanks for pointing this out!

Atharva-Kanherkar replied:
Hey @armanbilge!
I noticed the test is failing because we're only getting 6 successful polls instead of 10. Looking at the code more closely, I think I understand why: 

When we poll the queue, we get either:

- A singleton task (counts as 1 poll)
- An entire batch (also counts as 1 poll, even though it contains 10 tasks)
- Since we've added 5 singleton tasks and 1 batch to the queue, we're only able to poll 6 times successfully before the queue is empty.

Should I:

- Change the assertion to `assertEquals(polledCount, 6) ` to match the actual behavior?
- Change how we count polled items (counting each task in a batch separately)?
- Add more items to the queue so we can successfully poll 10 times?

armanbilge replied:
Oh, oops! Then the number should be 6 :) I thought it was 10 because of the loop counter (I guess that can be 6 too).

Atharva-Kanherkar replied:
Yes! I have updated it to 6. Let's hope everything works:)

Atharva-Kanherkar replied:
hey @armanbilge! all the checks pass:)

armanbilge commented on tests/jvm/src/test/scala/cats/effect/unsafe/ScalQueueSuite.scala:
```diff
@@ -0,0 +1,239 @@
+      if (element ne null) {
+        polledCount += 1
+
+        // Execute the polled task
```
Since all the tasks are no-op, we don't need to worry about executing them.

Atharva-Kanherkar replied:
Okay! Done!

Atharva-Kanherkar committed:
Update tests/jvm/src/test/scala/cats/effect/unsafe/ScalQueueSuite.scala

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

Atharva-Kanherkar committed:
Changes

Atharva-Kanherkar committed:
Added requested changes

armanbilge commented on build.sbt:
```diff
@@ -725,11 +725,16 @@ lazy val core = crossProject(JSPlatform, JVMPlatform, NativePlatform)
       // internal API change, makes CpuStarvationMetrics available on all platforms
       ProblemFilters.exclude[MissingClassProblem](
         "cats.effect.metrics.JvmCpuStarvationMetrics$NoOpCpuStarvationMetrics"),
+         // introduced by #4292 , external queue metrics
+// WorkStealingThreadPoolMetrics is a sealed trait, so we control all of its implementations.
+ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.effect.unsafe.metrics.WorkStealingThreadPoolMetrics.externalQueue"),
```
Can you run `sbt scalafmtSbt` to properly format this file? Normally CI is supposed to catch this, but I realized it's inadvertently broken.
- https://github.com/typelevel/cats-effect/pull/4373

Atharva-Kanherkar replied:
Okay sure! I did run scalafmtAll, I'll run this too.

Atharva-Kanherkar replied:
Btw, just curious, why is it broken?

armanbilge replied:
> Btw, just curious, why is it broken?

You can check the PR I linked for more details ... basically we ran the check at the wrong time, after selecting a specific sub-project for CI (e.g. JVM or JS or Native). So it's not running on the `build.sbt` for the whole project.

Atharva-Kanherkar replied:
Okay! Thanks for this! I have added the formatting.

Atharva-Kanherkar committed:
Formatting

Atharva-Kanherkar committed:
Added changes

armanbilge commented on tests/jvm/src/test/scala/cats/effect/unsafe/ScalQueueSuite.scala:
```diff
@@ -0,0 +1,216 @@
+    val batch = new Array[Runnable](batchSize)
+    var i = 0
+    while (i < batchSize) {
+      batch(i) = new Runnable { def run(): Unit = () }
```
This can also be `() => ()`, all of the places you did `new Runnable` can be short-handed.

Atharva-Kanherkar replied:
Okay for sure!

armanbilge commented on build.sbt:
```diff
@@ -725,11 +725,16 @@ lazy val core = crossProject(JSPlatform, JVMPlatform, NativePlatform)
       // internal API change, makes CpuStarvationMetrics available on all platforms
       ProblemFilters.exclude[MissingClassProblem](
         "cats.effect.metrics.JvmCpuStarvationMetrics$NoOpCpuStarvationMetrics"),
+      // introduced by #4292 , external queue metrics
+// WorkStealingThreadPoolMetrics is a sealed trait, so we control all of its implementations.
```
Is it just me, or does the formatting here still seem strange?

Atharva-Kanherkar replied:
I ran the commands, I think ill adjust it manually.

Atharva-Kanherkar replied:
hey @armanbilge does the formatting seem okay now? i just manually aligned it

Atharva-Kanherkar committed:
Added shorthands

armanbilge committed:
Formatting fixes

Atharva-Kanherkar commented:
hey @armanbilge ! do you need anything else from my side! :) happy to help..

armanbilge commented:
@djspiewak do you have any performance concerns here, should we run some benchmarks?

djspiewak commented:
I would definitely like to run some benchmarks on this one. I'll review the diff more closely though.

Atharva-Kanherkar commented:
Hi @armanbilge @djspiewak! Hope you're doing well. Just wanted to follow up on this—any updates by chance?

djspiewak committed:
Merge branch 'series/3.x' into external-queue-tracking

djspiewak commented:
Finally got to the benchmarks. Looks like there is a notable impact. Let's see if we can do some micro-optimization to gain back those millis.

## Before

```
[info] Benchmark                                             (cpuTokens)   (size)   Mode  Cnt    Score   Error    Units
[info] ParallelBenchmark.parTraverse                               10000     1000  thrpt   10  288.943 ± 1.853    ops/s
[info] ParallelBenchmark.traverse                                  10000     1000  thrpt   10   48.077 ± 0.019    ops/s
[info] WorkStealingBenchmark.alloc                                   N/A  1000000  thrpt   10    4.473 ± 0.033  ops/min
[info] WorkStealingBenchmark.manyThreadsSchedulingBenchmark          N/A  1000000  thrpt   10    6.820 ± 0.147  ops/min
[info] WorkStealingBenchmark.runnableScheduling                      N/A  1000000  thrpt   10  700.437 ± 3.968  ops/min
[info] WorkStealingBenchmark.runnableSchedulingScalaGlobal           N/A  1000000  thrpt   10  597.162 ± 0.807  ops/min
[info] WorkStealingBenchmark.scheduling                              N/A  1000000  thrpt   10    8.120 ± 0.346  ops/min

```

## After

```
[info] Benchmark                                             (cpuTokens)   (size)   Mode  Cnt    Score   Error    Units
[info] ParallelBenchmark.parTraverse                               10000     1000  thrpt   10  289.627 ± 1.154    ops/s
[info] ParallelBenchmark.traverse                                  10000     1000  thrpt   10   48.061 ± 0.012    ops/s
[info] WorkStealingBenchmark.alloc                                   N/A  1000000  thrpt   10    4.427 ± 0.009  ops/min
[info] WorkStealingBenchmark.manyThreadsSchedulingBenchmark          N/A  1000000  thrpt   10    6.713 ± 0.132  ops/min
[info] WorkStealingBenchmark.runnableScheduling                      N/A  1000000  thrpt   10  626.559 ± 3.414  ops/min
[info] WorkStealingBenchmark.runnableSchedulingScalaGlobal           N/A  1000000  thrpt   10  579.224 ± 2.218  ops/min
[info] WorkStealingBenchmark.scheduling                              N/A  1000000  thrpt   10    7.908 ± 0.388  ops/min
```

Atharva-Kanherkar commented:
@djspiewak thank you so much for taking your time to review this!

Atharva-Kanherkar commented:
> Finally got to the benchmarks. Looks like there is a notable impact. Let's see if we can do some micro-optimization to gain back those millis.
> 
> ## Before
> ```
> [info] Benchmark                                             (cpuTokens)   (size)   Mode  Cnt    Score   Error    Units
> [info] ParallelBenchmark.parTraverse                               10000     1000  thrpt   10  288.943 ± 1.853    ops/s
> [info] ParallelBenchmark.traverse                                  10000     1000  thrpt   10   48.077 ± 0.019    ops/s
> [info] WorkStealingBenchmark.alloc                                   N/A  1000000  thrpt   10    4.473 ± 0.033  ops/min
> [info] WorkStealingBenchmark.manyThreadsSchedulingBenchmark          N/A  1000000  thrpt   10    6.820 ± 0.147  ops/min
> [info] WorkStealingBenchmark.runnableScheduling                      N/A  1000000  thrpt   10  700.437 ± 3.968  ops/min
> [info] WorkStealingBenchmark.runnableSchedulingScalaGlobal           N/A  1000000  thrpt   10  597.162 ± 0.807  ops/min
> [info] WorkStealingBenchmark.scheduling                              N/A  1000000  thrpt   10    8.120 ± 0.346  ops/min
> ```
> 
> ## After
> ```
> [info] Benchmark                                             (cpuTokens)   (size)   Mode  Cnt    Score   Error    Units
> [info] ParallelBenchmark.parTraverse                               10000     1000  thrpt   10  289.627 ± 1.154    ops/s
> [info] ParallelBenchmark.traverse                                  10000     1000  thrpt   10   48.061 ± 0.012    ops/s
> [info] WorkStealingBenchmark.alloc                                   N/A  1000000  thrpt   10    4.427 ± 0.009  ops/min
> [info] WorkStealingBenchmark.manyThreadsSchedulingBenchmark          N/A  1000000  thrpt   10    6.713 ± 0.132  ops/min
> [info] WorkStealingBenchmark.runnableScheduling                      N/A  1000000  thrpt   10  626.559 ± 3.414  ops/min
> [info] WorkStealingBenchmark.runnableSchedulingScalaGlobal           N/A  1000000  thrpt   10  579.224 ± 2.218  ops/min
> [info] WorkStealingBenchmark.scheduling                              N/A  1000000  thrpt   10    7.908 ± 0.388  ops/min
> ```

i see. could you please elaborate a little bit on how we could do that?

djspiewak commented:
No magic bullet really! We just have to experiment and see. The most expensive things you added have to do with the `AtomicLong`s, so I would start by commenting out the mutations in the hot path and see if that changes the benchmark results.

Or rather, I would start by trying to replicate my benchmark results locally. `benchmarks / Jmh / run -f 1 -wi 10 -i 10 cats.effect.benchmarks.WorkStealingBenchmark` is the relevant one. See if you can reproduce the statistically significant difference before and after your change, then see if you can manipulate it by commenting out lines. If we can figure out what's especially expensive, we can start thinking of creative ways to do it differently so it's less expensive!
